scanner/decisions/0000-adopt-adrs.md

```diff
@@ -0,0 +1,37 @@
+# 0000 - Adopt ADRs
+
+- **Author(s):** J. Victor Martins <jvdm@sdf.org>
+- **Created:** [2023-08-23 Wed]
+
+## Status
+
+Accepted.
+
+## Context
+
+Currently, decisions about Stackrox Scanner are made using informal slack threads, ad-hoc discussions, and occasionally when the problem is considered complex, design documents. They are:
+
+- **Hard to Navigate:** Finding discussions, decisions, and their context can sometimes be challenging. There is no decision log providing context on the implementation and design information.
+- **Opportunistic:** Decisions might be made too fast without all relevant consideration.
+- **Overlooks Long-Term Impact:** Quick decisions might only partially evaluate the long-term effects on the project.
+- **Lacks a Well-Defined Lifecycle:** There's no structured review process to guarantee why a decision was accepted or rejected at the end.
+- **Unclear Boundaries:** There's no clear framework to help us understand when the overhead of design documents is worth it.
+
+## Decision
+
+1. **ADRs for long-term impact decisions:** ADRs will be used to document decisions that have a long-term impact on the project. This process is considered more mature and reliable than other methods (e.g., slack threads).  We will continue to use informal methods for trivial decisions.
+2. **ADRs for smaller and larger decisions:** ADRs can be used for smaller and larger decisions, whether public or not, depending on the context.
+3. **Design docs to aid complex contexts:** Design documents can be used when the context is ambiguous and requires an in-depth analysis of options available. ADRs will still be used to formally record the decision.[^1]
+4. **ADRs are public and local by default:** ADRs will be kept in a specific directory within the code repo to correlate decisions with code changes and make them widely available, even to external parties. The ADR will be stored in a private location when the context is sensitive (e.g., security, customer information).
+5. **ADR template:** ADRs will use a well-defined template.
```

Comment:
Will this template include where to store binary assets?  For example, if the ADR includes a diagram / picture I would expect anyone with edit access to the ADR to also be able to edit the diagrams/pictures/etc.  Or perhaps we settle on using only assets that can be represented 'as code', ie: diagrams via [mermaidjs](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/)

Reply:
I haven't considered that yet. I agree it's important to make sure diagrams are editable, but I'm not sure which platform would be best. Instead of enforcing, I believe preferring code-based diagrams would be better. Additionally, there is a restriction on the size of files in the repo (50K, I think), which will require us to use simple and focused diagrams. We can certainly improve things as we go and document the options in the `template.md` file.

Reply:
Thinking about this more, perhaps a link to lucid, miro, whatever if thats where the diagram was created would be sufficient.  Ideally placing them into a shared folder/group mirroring GH repo access.

That max size must be adjustable - In my past we've stored videos in GH that were 500 MiB+

Reply:
As long as anyone can access the link and see the diagram, I would be OK.

> That max size must be adjustable

FYI, I think that's part of our post-commit checks, but I am not sure.